sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1174 @@
+                "You can set the request value to RequestType.UNREQUESTED "
+                "to silence this warning, or to RequestType.REQUESTED to "
+                "consume and use the metadata."
+            )
+
+    def _route_params(self, params=None):
+        """Prepare the given parameters to be passed to the method.
+
+        The output of this method can be used directly as the input to the
+        corresponding method as extra props.
+
+        Parameters
+        ----------
+        params : dict
+            A dictionary of provided metadata.
+
+        Returns
+        -------
+        params : Bunch
+            A :class:`~utils.Bunch` of {prop: value} which can be given to the
+            corresponding method.
+        """
+        self._check_warnings(params=params)
+        params = {} if params is None else params
+        args = {arg: value for arg, value in params.items() if value is not None}
+        res = Bunch()
+        for prop, alias in self._requests.items():
+            if RequestType.is_valid(alias):
+                alias = RequestType(alias)
+
+            if alias == RequestType.UNREQUESTED or alias == RequestType.WARN:
+                continue
+            elif alias == RequestType.REQUESTED and prop in args:
+                res[prop] = args[prop]
+            elif alias == RequestType.ERROR_IF_PASSED and prop in args:
+                raise ValueError(
+                    f"{prop} is passed but is not explicitly set as "
+                    f"requested or not for {self.owner}.{self.method}"
+                )
+            elif alias in args:
+                res[prop] = args[alias]
+        return res
+
+    def _serialize(self):
+        """Serialize the object.
+
+        Returns
+        -------
+        obj : dict
+            A serialized version of the instance in the form of a dictionary.
+        """
+        return {
+            prop: RequestType(alias) if RequestType.is_valid(alias) else alias
+            for prop, alias in self._requests.items()
+        }
+
+    def __repr__(self):
+        return str(self._serialize())
+
+    def __str__(self):
+        return str(repr(self))
+
+
+class MetadataRequest:
+    """Contains the metadata request info of a consumer.
+
+    Instances of :class:`MethodMetadataRequest` are used in this class for each
+    available method under `metadatarequest.{method}`.
+
+    Consumer-only classes such as simple estimators return a serialized
+    version of this class as the output of `get_metadata_routing()`.
+
+    .. versionadded:: 1.1
+
+    Parameters
+    ----------
+    owner : str
+        The name of the object to which these reqeusts belong.
+    """
+
+    # this is here for us to use this attribute's value instead of doing
+    # `isinstance`` in our checks, so that we avoid issues when people vendor
+    # this file instad of using it directly from scikit-learn.
+    type = "request"
```

Comment:
I meant the attribute name `type`, not its values like `"request"`. As this attribute is not inherited, e.g. via `_MetadataRequester`, it seems fine. I was worried about name clashes.

Reply:
ah I see. Yeah I wouldn't want to have two different names for it since this way I can just check for the value of the variable w/o having to care which instance type it is. I think it should be private anyway, so making it private.